packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,247 @@
+                    fastWithdrawal.amount,
+                    fastWithdrawal.feeToken,
+                    fastWithdrawal.fastFee,
+                    fastWithdrawal.fee,
+                    fastWithdrawal.onchainFeePayment,
+                    fastWithdrawal.fastUntil,
+                    fastWithdrawal.validUntil
+                )
+            )
+        );
+
+        // Make sure the tx isn't used multiple times
+        require(txHash[hash] == false, "TX_REPLAY");
+        txHash[hash] = true;
+
+        // Check the signature
+        require(hash.verifySignature(fastWithdrawal.from, fastWithdrawal.signature), "INVALID_SIGNATURE");
+
+        // The liquidity provider always authorizes the fast withdrawal by being the direct caller
+        address payable liquidityProvider = msg.sender;
+
+        IExchangeV3 exchange = IExchangeV3(fastWithdrawal.exchange);
+
+        uint fee;
+        if (now <= fastWithdrawal.fastUntil) {
+            /* Fast withdrawal */
+
+            // Transfer the tokens immediately to the requested address
+            // using funds from the liquidity provider (`msg.sender`).
+            transfer(
+                liquidityProvider,
+                fastWithdrawal.to,
+                fastWithdrawal.token,
+                fastWithdrawal.amount
+            );
+
+            // Pay the fast fee
+            fee = fastWithdrawal.fastFee;
+
+            // If the fee is paid offchain in the same token as the main transfer
+            // just add it to the transfer amount
+            if (!fastWithdrawal.onchainFeePayment && fastWithdrawal.token == fastWithdrawal.feeToken) {
+                fastWithdrawal.amount = fastWithdrawal.amount.add(fee);
+                fee = 0;
+            }
+
+            // Approve the offchain transfer from the account that's withdrawing back to the liquidity provider
+            exchange.approveOffchainTransfer(
+                fastWithdrawal.from,
+                liquidityProvider,
+                fastWithdrawal.token,
+                fastWithdrawal.amount
+            );
+        } else {
+            /* Standard onchain withdrawal */
+
+            // Check the time limit
+            require(now <= fastWithdrawal.validUntil, "TX_EXPIRED");
+
+            // Do the onchain withdrawal request
+            (,,, uint withdrawalFeeETH) = exchange.getFees();
+            exchange.withdraw{value: withdrawalFeeETH}(
+                fastWithdrawal.from,
+                fastWithdrawal.token,
+                uint96(fastWithdrawal.amount)
+            );
+
+            // Pay the standard fee
+            fee = fastWithdrawal.fee;
+        }
+
+        // Fee payment to the liquidity provider
+        if (fee > 0) {
+            if (fastWithdrawal.onchainFeePayment) {
+                // Do fee payment directly from the user's wallet if requested
+                // (using the approval for the exchange)
+                exchange.onchainTransferFrom(
```

Comment:
should we check the following in the beginning fo this function
```
require(!onchainFeePayment || fastWithdrawal.feeToken!=address(0))
```

or if `feeToken==address(0)` we force to use get fees from offchain?

Reply:
I don't think this needs to be checked because the fast withdrawal won't be executed if no fee would get paid. With the current code it actually already throws because `onchainTransferFrom` will try to transfer using ERC20(feeToken). These kind of things can be checked offchain, nobody would want to pay the gas to get an invalid request onchain to begin with.